packages/docutils/lib/builder/nav.ts

```diff
@@ -0,0 +1,320 @@
+ * @param mkDocsDocsDir The path to the `docs_dir` in the `mkdocs.yml` file; must be absolute
+ * @returns List of Markdown files relative to the `docs_dir` in the `mkdocs.yml` file
+ */
+async function findRelativeMarkdownFiles(
+  targetDir: string,
+  mkDocsDocsDir: string
+): Promise<string[]> {
+  if (!path.isAbsolute(targetDir)) {
+    throw new DocutilsError(`Expected absolute path, got '${targetDir}'`);
+  }
+  if (!path.isAbsolute(mkDocsDocsDir)) {
+    throw new DocutilsError(`Expected absolute path, got '${mkDocsDocsDir}'`);
+  }
+  const relDir = path.relative(mkDocsDocsDir, targetDir);
+  const dirEnts = await fs.readdir(targetDir, {withFileTypes: true});
+  return dirEnts
+    .filter((ent) => ent.isFile() && ent.name.endsWith('.md'))
+    .map((ent) => path.join(relDir, ent.name));
+}
+
+/**
+ * Because the `nav` property of `mkdocs.yml` is both a recursive type and a kind of awful one, it's
+ * easier to work with it if we rewrite the data into a flat array of objects. We keep a `keypath`
+ * prop which represents the deep/nested location within the `nav` object.
+ *
+ * @privateRemarks This function is not recursive; instead it loops over a queue of items to process
+ * data, and we append to that queue while processing if needed.
+ * @param nav Contents of the `nav` prop of `mkdocs.yml`
+ * @returns A list of objects, each with a `keypath` property and a `fileOrUrl` property (and maybe
+ * a `name` property)
+ */
+export function parseNav(nav: MkDocsYmlNav): ParsedNavData[] {
+  let parsedNav: ParsedNavData[] = [];
+  const entries = Object.entries(nav);
+  type QueueItem = {
+    entries: typeof entries;
+    keypath: string;
+  };
+  const queue: QueueItem[] = [{entries, keypath: ''}];
+
+  while (queue.length) {
+    const {entries, keypath} = queue.shift()!;
+    for (const [key, item] of entries) {
+      if (_.isString(item)) {
+        const navData: ParsedNavData = {
+          keypath: keypath ? `${keypath}.${key}` : key,
+          fileOrUrl: item,
+        };
+
+        // if the key is not convertible to a number, it's a name
+        // which was manually put there by somebody.
+        if (isNaN(Number(key))) {
+          navData.name = key;
+        }
+        parsedNav = [...parsedNav, navData];
+      } else if (_.isObject(item)) {
+        const subEntries = Object.entries(item);
+        queue.push({entries: subEntries, keypath: keypath ? `${keypath}.${key}` : key});
+      }
+    }
+  }
+  return parsedNav;
+}
+
+/**
+ * Finds all items within the list of parsed nav data which correpsond to the header.  This is
+ * imperfect, as it's possible for the header string to appear in multiple places in the nav, but let's just ignore that until we can't.
+ * @param navData Some parsed nav data
+ * @param header Header name
+ */
+function itemsForHeader(navData: ParsedNavData[], header: string) {
+  return _.filter(navData, (item) => _.toPath(item.keypath).includes(header));
```

Comment:
can keypath be null here?

Reply:
no, `keypath` cannot be null; see the type of `ParsedNavData`